django/db/models/query.py

```diff
@@ -942,6 +942,8 @@ def extra(self, select=None, where=None, params=None, tables=None,
 
     def reverse(self):
         """Reverse the ordering of the QuerySet."""
+        if not self.query.can_filter():
+            raise TypeError('Cannot reverse a query once a slice has been taken')
```

Comment:
That's what I wrote in an outdated diff regarding this `raise TypeError`:

> I'm not sure about `TypeError` either.  I tried to stay close to similar exceptions like
> - [`raise TypeError("Cannot call delete() after .values() or .values_list()")`](https://github.com/django/django/blob/29f6079/django/db/models/query.py#L594)
> - [`raise TypeError("Cannot call select_related() after .values() or .values_list()")`](https://github.com/django/django/blob/29f6079/django/db/models/query.py#L856)
> - [`raise TypeError("Cannot call defer() after .values() or .values_list()")`](https://github.com/django/django/blob/29f6079/django/db/models/query.py#L975)
> - [`raise TypeError("Cannot call only() after .values() or .values_list()")`](https://github.com/django/django/blob/29f6079/django/db/models/query.py#L990)

It would be nice to unify the errors.